pkg/daemon/bootc.go

```diff
@@ -0,0 +1,222 @@
+func NewClient(id string) Client {
+	return Client{
+		clientid: id,
+	}
+}
+
+func (client *Client) newCmd(args ...string) *exec.Cmd {
+	r := exec.Command("bootc", args...)
+	r.Env = append(r.Env, "BOOTC_CLIENT_ID", client.clientid)
+	return r
+}
+
+func (client *Client) run(args ...string) error {
+	c := client.newCmd(args...)
+	return c.Run()
+}
+
+// QueryStatus loads the current system state.
+func (client *Client) QueryStatus() (*BootcStatus, error) {
+	var q BootcStatus
+	c := client.newCmd("status", "--json")
+	buf, err := c.Output()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(buf, &q); err != nil {
+		return nil, fmt.Errorf("failed to parse `bootc status --json` output: %w", err)
+	}
+
+	return &q, nil
+}
+
+// GetBootedImage finds the booted image, or returns nil if none is found.
+func (s *Status) GetBootedImage() *BootEntry {
+	return s.Booted
+}
+
+// GetStagedImage finds the staged image for the next boot, or returns nil if none is found.
+func (s *Status) GetStagedImage() *BootEntry {
+	return s.Staged
+}
+
+// GetRollbackImage finds the rollback deployment, or returns nil if none is found.
+func (s *Status) GetRollbackImage() *BootEntry {
+	return s.Rollback
+}
+
+// GetBaseChecksum returns the ostree commit used as a base.
+func (b *BootEntry) GetBaseChecksum() string {
+	return b.Ostree.Checksum
+}
+
+// GetContainerImageReference returns the the container image reference.
+func (b *BootEntry) GetContainerImageReference() string {
+	return b.Image.Image.Image
+}
+
+// imgref.go
+// ImageReference captures an image signature verification policy alongside an image reference.
+// This mirrors https://github.com/containers/bootc/blob/94521ecb19b8d0f2bfc36220322c60cafa036296/lib/src/spec.rs#L69
+type ImageReference struct {
+	Image     string          `json:"image"`
+	Transport string          `json:"transport"`
+	Signature *ImageSignature `json:"signature,omitempty"`
+}
+
+// ImageSignature mirrors https://github.com/containers/bootc/blob/94521ecb19b8d0f2bfc36220322c60cafa036296/lib/src/spec.rs#L57
+// Same as https://docs.rs/ostree-ext/latest/ostree_ext/container/enum.SignatureSource.html
+type ImageSignature struct {
+	AllowInsecure bool
+	OstreeRemote  string
+}
+
+// bootc.go
+
+// BootcClient provides all Bootc related methods in one structure.
+type BootcClient struct {
+	client Client
+}
+
+// Synchronously invoke bootc, writing its stdout to our stdout,
+// and gathering stderr into a buffer which will be returned in err
+// in case of error.
+func runBootc(args ...string) error {
+	return runCmdSync("bootc", args...)
+}
+
+func (b *BootcClient) Initialize() error {
+	// Commands like update and rebase need the pull secrets to pull images and manifests,
+	// make sure we get access to them when we Initialize
+	err := useMergedPullSecrets(bootcSystem)
+	if err != nil {
+		klog.Errorf("error while linking bootc pull secrets %v", err)
```

Comment:
question: do we just want to log the error and continue when an error occurs here? If not we should do `return useMergePullSecrets()` to ensure we actually error out when an error happens.

Reply:
Let's return the error and cut it there. I was following this check that currently exists in rpm-ostree:
```
// Commands like update and rebase need the pull secrets to pull images and manifests,
	// make sure we get access to them when we Initialize
	err := useMergedPullSecrets(rpmOstreeSystem)
	if err != nil {
		klog.Errorf("error while linking rpm-ostree pull secrets %v", err)
	}
```
But with a second thought, I think it is something to be prevented.

Reply:
I was wondering the same thing too, we should confirm this as it is better to completely error out than just log and wait for it to fail at a later stage.

Reply:
I was wondering the same thing too, we should confirm this as it is better to completely error out than just log and wait for it to fail at a later stage.

Reply:
I think at least  in the following implications
- Switch 
- RebaseLayered 
- RebaseLayeredFromContainerStorage
We def need the presence and correct reading of the pull secret.

Reply:
I have error-ed them out in the latest changes according to discussions here:
[1] Keeping the error will result in fail at a later time and hard-to-track bugs: https://github.com/openshift/machine-config-operator/pull/4465#discussion_r1679497966
https://github.com/openshift/machine-config-operator/pull/4465#discussion_r1681257687
[2] Several frequently invoked rpm-ostree/bootc commands need the presence of the pull secret. Better to set them up at initialization: https://github.com/openshift/machine-config-operator/pull/4465#discussion_r1681264154

Reply:
thought: Maybe do this change only for the bootc code path and open another PR for fixing this in the rpm-ostree code path as that is actually affecting code actively used today. This way we can test it separately and avoid any potential regressions. What do you think?

Reply:
Update:
Kept for rpm-ostree initialization only: 
```
	err := useMergedPullSecrets(...)
	if err != nil {
		klog.Errorf("error while linking rpm-ostree pull secrets %v", err)
	}
```
Updated to for bootc switch and rpm-ostree rebase because the pull secret is mandatory for future steps
```
	err := useMergedPullSecrets(...)
	if err != nil {
		return fmt.Errorf("Error while ensuring access to pull secrets: %w", err)
	}
```